nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/ExtractEmailAttachments.java

```diff
@@ -0,0 +1,201 @@
+            .name("original")
+            .description("The original file")
+            .build();
+    public static final Relationship REL_FAILURE = new Relationship.Builder()
+            .name("failure")
+            .description("Flowfiles that could not be parsed")
+            .build();
+    private Set<Relationship> relationships;
+    private List<PropertyDescriptor> descriptors;
+
+
+    @Override
+    protected void init(final ProcessorInitializationContext context) {
+        final Set<Relationship> relationships = new HashSet<>();
+        relationships.add(REL_ATTACHMENTS);
+        relationships.add(REL_ORIGINAL);
+        relationships.add(REL_FAILURE);
+        this.relationships = Collections.unmodifiableSet(relationships);
+
+        final List<PropertyDescriptor> descriptors = new ArrayList<>();
+
+        this.descriptors = Collections.unmodifiableList(descriptors);
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) {
+        final ComponentLog logger = getLogger();
+        final FlowFile originalFlowFile = session.get();
+        if (originalFlowFile == null) {
+            return;
+        }
+        final List<FlowFile> attachmentsList = new ArrayList<>();
+        final List<FlowFile> invalidFlowFilesList = new ArrayList<>();
+        final List<FlowFile> originalFlowFilesList = new ArrayList<>();
+
+        session.read(originalFlowFile, new InputStreamCallback() {
+                @Override
+                public void process(final InputStream rawIn) throws IOException {
+                    try (final InputStream in = new BufferedInputStream(rawIn)) {
+                        Properties props = new Properties();
+                        Session mailSession = Session.getDefaultInstance(props, null);
+                        MimeMessage originalMessage = new MimeMessage(mailSession, in);
+                        MimeMessageParser parser = new MimeMessageParser(originalMessage).parse();
+                        // RFC-2822 determines that a message must have a "From:" header
+                        // if a message lacks the field, it is flagged as invalid
+                        Address[] from = originalMessage.getFrom();
+                        Date sentDate = originalMessage.getSentDate();
+                        if (from == null || sentDate == null ) {
+                            // See ya later my friend...
+                            throw new MessagingException("Message failed RFC2822 validation");
+                        }
+                        if (parser.hasAttachments()) {
+                            final String originalFlowFileName = originalFlowFile.getAttribute(CoreAttributes.FILENAME.key());
+                            for (final DataSource data : parser.getAttachmentList()) {
+                                FlowFile split = session.create(originalFlowFile);
```

Comment:
If an error occurs in this block (after a split has been created) this error[1] will get thrown. This is because all FlowFiles need to be handled even if they were just created. What is probably best is to have another try/catch which covers the contents of this for loop and removes the newly created split flowfile if it fails. 

Also this would allow for potentially extracting some splits and fail to split others. This may require a property to determine what to do. Of course, in the event of a failure you could just remove all the splits and transfer the original to failure.

[1] FlowFileHandlingException: Cannot commit session because the following FlowFiles have not been removed or transferred: ...